### PR TITLE
Edit inline diff

### PR DESCRIPTION
**Pull request by stoivo**
Now it work nice with:
 - staging\unstage line alone
 - staging\unstage line in middle of hunk
 - undo staging\unstage for all action on single line
 - undo for revert hunk

This have been hard. But now it should be solved :)

### CODE REVIEW

File: common/util/debug.py [hunk 1]
```diff
@@ -53,3 +53,18 @@ def wrapped_fn(*args, **kwargs):
...
+
+def pprint(*args):
+    """
+    Pretty print since we can not use debugger
+    """
+    import pprint
+
+    # pp = pprint.PrettyPrinter(indent=4)
+    for arg in args:
+        if arg.__class__.__name__ == 'str':
```
Comment (divmain) [inline common/util/debug.py]: `if isinstance(arg, str):`

File: core/commands/inline_diff.py [hunk 2]
```diff
@@ -414,7 +415,7 @@ def run_async(self, reset=False):
         # 2) The user is in non-cached mode and wants to undo a line/hunk, so
         #    DO apply the patch in reverse, and do apply it both against the
         #    index and the working tree.
-        # 3) The user is in cached mode and wants to undo a line hunk, so DO
+        # 3) The user is in cached mode and wants to undo a line/hunk, so DO
```
Comment (divmain) [inline core/commands/inline_diff.py]: Thanks :)

File: core/git_command.py [hunk 3]
```diff
@@ -60,9 +60,10 @@ def git(self, *args, stdin=None, working_dir=None, show_panel=False, throw_on_st
         of the git command as a string.
 
         If stdin is provided, it should be a string and will be piped to
-        the git process.  If `working_dir` is provided, set this as the
-        current working directory for the git process; otherwise,
-        the `repo_path` value will be used.
```
Comment (divmain) [inline core/git_command.py]: Can you please revert this change?

**Comment by divmain** [comment]
Wow, this is complex.  Let me pull it down tomorrow and take a look.  Want to make sure this doesn't introduce any new bugs, and addresses the issue.  Really appreciate your work on this, and I'll let you know soon.

File: core/commands/inline_diff.py [hunk 4]
```diff
@@ -471,25 +488,52 @@ def get_diff_from_line(self, line_no, reset):
...
         line_type = hunk_ref.line_types[index_in_hunk]
 
+        # need to make undo work properly when undoing
+        # a specific line.
+        for type in hunk_ref.line_types[:index_in_hunk]:
+            if type == "-":
+                cur_hunk_begin_on_minus += 1
+            else:
+                # type will be +
+                cur_hunk_begin_on_pluss += 1
```
Comment (divmain) [inline core/commands/inline_diff.py]: Can you change this to `cur_hunk_begin_on_plus` everywhere?

File: core/commands/inline_diff.py [hunk 5]
```diff
@@ -471,25 +488,52 @@ def get_diff_from_line(self, line_no, reset):
...
+            xhead_start = head_start - index_in_hunk + (0 if line_type == "+" else add_length_earlier_in_diff)
+
+            return ("@@ -{head_start},{head_length} +{new_start},{new_length} @@\n"
+                "{line_type}{line}").format(
+                head_start=(xhead_start if xhead_start >= 0 else cur_hunk_begin_on_pluss),
+                head_length="0" if line_type == "+" else "1",
+                # If head_length is zero, diff will report original start position
+                # as one less than where the content is inserted, for example:
+                #   @@ -75,0 +76,3 @@
+                # new_start=head_start + (0 if line_type == "+" else 1),index_in_hunk
```
Comment (divmain) [inline core/commands/inline_diff.py]: Remove this line, methinks?

File: core/commands/inline_diff.py [hunk 6]
```diff
@@ -471,25 +488,52 @@ def get_diff_from_line(self, line_no, reset):
...
+                new_start=head_start - 1 - cur_hunk_begin_on_minus + index_in_hunk + add_length_earlier_in_diff + (1 if line_type == "+" else 0),
+                # new_start=head_start + (-1 if line_type == "+" else 0),
```
Comment (divmain) [inline core/commands/inline_diff.py]: And this one too.

File: core/commands/inline_diff.py [hunk 7]
```diff
@@ -471,25 +488,52 @@ def get_diff_from_line(self, line_no, reset):
...
+                new_length="1" if line_type == "+" else "0",
+                line_type=line_type,
+                line=line
+            )
+
+        else:
+            # TODO: Investigate this off-by-one ???
```
Comment (divmain) [inline core/commands/inline_diff.py]: Is this relevant anymore?  Is it explained by your investigations here?

Comment (stoivo) [inline core/commands/inline_diff.py]: I think it is correct now since all is working :)

File: core/commands/inline_diff.py (hunk 6 above)
Comment (stoivo) [inline core/commands/inline_diff.py]: I would say no, just take it a away.

**Comment by stoivo** [comment]
I did the changes you requested and rebased and push :sparkles:

**Comment by divmain** [comment]
Fantastic!  I'll be working with this a bit more before I do the next release, but this is looking solid.  Thanks again for your work!
